File: python/dart_cuda/core/transformers/standard/transformer_decoder_block.py
"""Decoder transformer block — port of
`lib/transformer/transformer_decoder_block.dart` (dart-torch).

Three sub-layers (pre-norm):
    1. Masked multi-head self-attention.
    2. Multi-head cross-attention over encoder output.
    3. Position-wise feed-forward network.
"""

from __future__ import annotations

from typing import List

from ...attention.multi_head_attention import MultiHeadAttention
from ...attention.multi_head_cross_attention import MultiHeadCrossAttention
from ...layers.feed_forward import FeedForward
from ...layers.layer_norm import LayerNorm
from ...layers.nn import Module
from ...tensor.gpu_tensor import Tensor


class TransformerDecoderBlock(Module):
    def __init__(self, embed_size: int, num_heads: int, encoder_embed_size: int):
        self.embed_size = embed_size
        self.encoder_embed_size = encoder_embed_size
        self.self_attention = MultiHeadAttention(num_heads, embed_size, masked=True)
        self.cross_attention = MultiHeadCrossAttention(
            num_heads, embed_size, encoder_embed_size
        )
        self.ffn = FeedForward(embed_size)
        self.ln1 = LayerNorm(embed_size)
        self.ln2 = LayerNorm(embed_size)
        self.ln3 = LayerNorm(embed_size)

    # dart-torch aliases
    @property
    def selfAttention(self) -> MultiHeadAttention: return self.self_attention
    @property
    def crossAttention(self) -> MultiHeadCrossAttention: return self.cross_attention

    def forward(
        self,
        x_decoder: Tensor,
        x_encoder: Tensor,
        tracker: List[Tensor],
    ) -> Tensor:
        # 1. Masked self-attention.
        x_norm1 = self.ln1.forward(x_decoder, tracker)
        self_out = self.self_attention.forward(x_norm1, tracker)
        x_res1 = x_decoder + self_out
        tracker.extend([self_out, x_res1])

        # 2. Cross-attention over encoder output.
        x_norm2 = self.ln2.forward(x_res1, tracker)
        cross_out = self.cross_attention.forward(x_norm2, x_encoder, tracker)
        x_res2 = x_res1 + cross_out
        tracker.extend([cross_out, x_res2])

        # 3. FFN.
        x_norm3 = self.ln3.forward(x_res2, tracker)
        ffn_out = self.ffn.forward(x_norm3, tracker)
        out = x_res2 + ffn_out
        tracker.extend([ffn_out, out])
        return out

    def parameters(self) -> List[Tensor]:
        return [
            *self.self_attention.parameters(),
            *self.cross_attention.parameters(),
            *self.ffn.parameters(),
            *self.ln1.parameters(),
            *self.ln2.parameters(),
            *self.ln3.parameters(),
        ]

File: lib/core/transformers/standard/transformer_decoder_block.dart
/// Decoder transformer block with three pre-norm sub-layers:
///   1. Masked multi-head self-attention.
///   2. Multi-head cross-attention over encoder output.
///   3. Position-wise feed-forward network.
import 'package:dart_cuda/core/attention/multi_head_attention.dart';
import 'package:dart_cuda/core/attention/multi_head_cross_attention.dart';
import 'package:dart_cuda/core/layers/feed_forward.dart';
import 'package:dart_cuda/core/layers/layer_norm.dart';
import 'package:dart_cuda/core/layers/nn.dart';
import 'package:dart_cuda/core/tensor/gpu_tensor.dart';

class TransformerDecoderBlock extends Module {
  final int embedSize;
  final int encoderEmbedSize;
  final MultiHeadAttention selfAttention;
  final MultiHeadCrossAttention crossAttention;
  final FeedForward ffn;
  final LayerNorm ln1;
  final LayerNorm ln2;
  final LayerNorm ln3;

  TransformerDecoderBlock(this.embedSize, int numHeads, this.encoderEmbedSize)
    : selfAttention = MultiHeadAttention(numHeads, embedSize, masked: true),
      crossAttention = MultiHeadCrossAttention(
        numHeads,
        embedSize,
        encoderEmbedSize,
      ),
      ffn = FeedForward(embedSize),
      ln1 = LayerNorm(embedSize),
      ln2 = LayerNorm(embedSize),
      ln3 = LayerNorm(embedSize);

  Tensor forward(Tensor xDecoder, Tensor xEncoder, List<Tensor> tracker) {
    // 1. Masked self-attention.
    final xNorm1 = ln1.forward(xDecoder, tracker);
    final selfOut = selfAttention.forward(xNorm1, tracker);
    final xRes1 = xDecoder + selfOut;
    tracker.addAll([selfOut, xRes1]);

    // 2. Cross-attention over encoder output.
    final xNorm2 = ln2.forward(xRes1, tracker);
    final crossOut = crossAttention.forward(xNorm2, xEncoder, tracker);
    final xRes2 = xRes1 + crossOut;
    tracker.addAll([crossOut, xRes2]);

    // 3. FFN.
    final xNorm3 = ln3.forward(xRes2, tracker);
    final ffnOut = ffn.forward(xNorm3, tracker);
    final out = xRes2 + ffnOut;
    tracker.add(ffnOut);
    return out;
  }

  @override
  List<Tensor> parameters() => [
    ...selfAttention.parameters(),
    ...crossAttention.parameters(),
    ...ffn.parameters(),
    ...ln1.parameters(),
    ...ln2.parameters(),
    ...ln3.parameters(),
  ];
}
